### ISSUE TITLE
Lost JSON serialising ability with frozendict==2.2.0

### ISSUE THREAD

**Issue by zifeo** [Marco-Sulla/python-frozendict#47]
in 2.1.3: `orjson.dumps(frozendict.frozendict({"A": "B"})) == "{\"A\":\"B\"}"`
in 2.2.0: `frozendict.frozendict is not serializable`

This is not a real bug, but we lost something that was coming for free with `MutableMapping`.
What's the recommended workaround?

**Comment by Marco-Sulla** [comment]
The simpler workaround is to do:

    orjson.dumps(dict(frozendict.frozendict({"A": "B"})))

Anyway it's quite strange, since `frozendict` is registered as a `Mapping`. I have to investigate.

**Comment by Marco-Sulla** [comment]
## TL;DR
If you can, use `pickle` instead.

## Long explaining
It seems that this is a limitation of JSON in Python. Indeed I tried to serialize a MutableMapping using the builtin `json` module, and I get the same error.

The fact is that, before 2.2.0, the package released on Pypi was only the pure Python package. From 2.2.0 the C Extension is the default.

In the pure py package, `frozendict` inherits directly from `dict`. This is a weird trick to make it faster and more immutable. This is why JSON encoding was "supported".

The original version of Lezica did not inherit from `dict`. It inherited from `Mapping`, as the C Extension does. This is the way and I don't think it will change.

The main problem with `json` is that you can't register a custom serializer for your custom type, as you can do on the contrary with `pickle`. You must convert the custom type to a type supported by `json`, or create a custom serializer function or class and pass it to `json.dumps()`.

Furthermore, you can notice that `frozendict` is serialized as a `dict`. If you load the dump, you get a `dict`, not a `frozendict`.

I think I can't really fix anything here. Furthermore, JSON official support was never a goal for `frozendict`.
I suggested to use `pickle` if you can trust the `frozendict` data, or use the workaround in the previous comment.

If someone has a smarter idea, (s)he is welcome :)

**Comment by zifeo** [comment]
Thanks for the insider info! Pickle is not an option as I am not deserialising in Python. I only need a hashable dict class for building a recursive data structures, which might not be the primary goal of this lib. Using `orjson` and the dict constructor should do the job even if not the best solution.

**Comment by RhinocerosUnicornis** [comment]
> What's the recommended workaround?

You can use a custom serializer.

```py
import json
from frozendict import frozendict

class JSONResultEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, frozendict):
            return dict(o)

        return json.JSONEncoder.default(self, o)

def json_encoder(value):
    return json.dumps(value, cls=JSONResultEncoder)

as_json_string = json_encoder(frozendict({'foo': 'bar'}))
print(as_json_string)
```

**Comment by Marco-Sulla** [comment]
Well, I just committed a monkeypatch for `json` and `orjson`, so `frozendict` can be serialized to JSON by default even with the C Extension.

I have to say I don't like monkeypatches. I applied them in this case for a matter of usability. I hope in a future I can safely remove them. 

They are under the submodule `monkeypatch` and there are also functions for disabling them.

Let me know if they work and what do you think about.

**Comment by Marco-Sulla** [comment]
Well, I close this issue. Feel free to reopen it if something does not work (and since it's a monkeypatch, chances are high :D )
